### DISCUSSION TITLE
Parallel execution rules for "Actions"

### DISCUSSION THREAD

**Discussion by ahmedkhalaf** [modelint/shlaer-mellor-metamodel#57] [Q&A]
I was wondering what are the rules governing concurrent execution of "Actions", especially actions that involve other "instances".

Except for "run to completion" and "leave the system consistent" I didn't find much more in the backmatter chapter of "[Models to Code: With No Mysterious Gaps](https://learning.oreilly.com/highlights/9781484222164/)".
There's a rule that states: "Only one state activity of a given instance can be in execution at any time".
Also, I don't believe "run to completion" has the same meaning as a "transaction" where everything read/written within the action belongs to a consistent snapshot of the system.
So, I believe this might not be sufficient for correct model execution if "Multiple actions in an activity can execute concurrently" regardless of data dependencies (RAW constraint).

Is there a policy that helps avoid situations where actions might yield incorrect/inconsistent results?
For example:
1- An instance of class A is executing a "Qualification" or "Filtering" / "select" action on instances of class B based on the value of attribute B.X.
2- Concurrently, a few instances of class B execute an action potentially updating the value of their own attribute B.X.

I appreciate any hints or pointers if I'm missing something here.

Note: This issue might be inherited from "Procedures" which seems to be the broader term used in the "Rules about Procedures" section in "Executable UML: A Foundation for Model-Driven Architecture"

**Comment by mangoa01** [comment]
Leon asked that I make a few comments here.

You are correct that run to completion of state actions is not sufficient to insure a transaction on the data model. The usual rule is that a state action must "leave the data model consistent or generate an event which will eventually bring the data model into a consistent state." I interpret this rule in the model execution domains I have written as defining a "thread of control" and that thread of control is a transaction on the data model. So, for example, an event which arises from outside of a domain that initiates a state transition starts a thread of control that includes the action directly associated with the transition and all actions that execute as a result of event dispatches which are started should the original action signal any events. At the end of the thread of control, the data model must be in a consistent  referential state.

How accesses to the data model are handled in the face of concurrent execution is the purview of the model execution domain. For simpler mappings of a model onto a single threaded execution architecture, there are no data races by definition. The transaction implied by the thread of control exists to enforce referential integrity. Concurrent execution of multiple threads of control must be managed by the model execution domain (if is supports such functionality). This is no different than the problem faced by Relational Database Management Systems in their guarantees of serializable transactions and there are a number of implementation strategies to insure that each transaction appears to run in an isolated fashion. Clearly a more difficult problem for a model execution domain to solve.

In the end, these types of matters are not solved by the modelling language, i.e. one does not "model" the concurrency issues in a domain model. The modeling rules are purposely stated to allow a range of possible implementations. Rather, one chooses an appropriate model execution domain with the capabilities necessary to make efficient use of the underlying computing resources. Your example is a classic "data race" and whatever means by which the program is constructed, it will have to account for that data race to be correct. Such is the ugly aspect of a multi-threaded, shared-state execution environments (and one of the reasons it is so extraordinarily difficult to make correct programs in those environments). In our case the accounting is not done in the description of the application logic but is handled by the model execution domain for all classes of applications you choose to run on a given model execution domain.

Andrew Mangogna

**Reply by ahmedkhalaf** [reply]
Thank you so much for the detailed answer.
I understand that the MX domain will manage the execution in parallel threads (or otherwise) and that the concrete platform will determine the parallelism available for the MX implementation.
My understanding is that all execution rules are invariably specified and must be fulfilled by any MX domain on all platforms.

I believe there are two main issues here.
One main issue is about the model semantics, especially with filter actions even if the platform ensures mutual execution of actions and prevents the data race.
How does one model these distinct cases:
1- Selection is a separate action from subsequent actions. It doesn't matter if the instance attributes (incl. the ones used as filter criteria) change later.
2- Filter action is meant to be implemented as a condition guarding another action (while iterating over a set of instances).
3- What about guarding a sequence of actions?

The other main issue is: Isn't MX domain supposed to fulfill the model execution (correctness) rules on the target platform (for optimal performance) at an abstract level for any domain model?
I'm a bit confused, because I don't see how this is dependent on the implementation technology platform.

Also, There are already several model execution rules that are implementation-agnostic, apply to all application classes and still sometimes ordering/atomicity is clearly stated.

For example, the following rules from "Executable UML: A Foundation for Model-Driven Architecture" dictate ordering, atomicity, and priority.
Combined, they seem to work perfectly (protecting from data race conditions) as long as the instances don't access each other's data (except via copying values into signals):
- Once initiated, a procedure of an object must complete before another signal can be accepted by the same object. It is the modeler's responsibility to ensure that the procedure will complete.
- A procedure must leave data describing its own instance consistently. If a procedure updates an attribute of its own instance, it must update all attributes that are derived from the first attribute.
- If a procedure creates or deletes instances of its own class, it must ensure that any links involving those instances are made consistent with the rules stated on the class diagram (by action or by signal).
- When a procedure completes, it must leave the system consistent, either by writing data (described in the three rules above) or by generating signals to cause other objects to come into conformance with the data changes made by the sender of the signal.
- Signals are never lost: Every signal will be delivered to the object or external entity to which it is directed.
and
- Signals sent to self are always accepted before other signals to that instance.

These will be translated into an implementation that guarantees prioritization of self-directed signals and reliable delivery of signals somehow, but the implementation details (for example using POSIX threads and locks or some other constructs) are up to the implementation.

Similarly, I think there might be some room for additional rules (applicable regardless of the modeled application).
For example changing this rule as follows might resolve the issue:
- Multiple accessors of an object may execute concurrently, with respect to each other and to state procedures +[given that all attributes read by one action are not written by another].
+ Implicitly this means an action has to wait if it would update an attribute within any of the instances accessed by "filter" action for example.

Please note that there are several ways to achieve that by the implementation, for example:
- an embedded system might opt for mutual execution among all actions to maximize efficiency
- an HPC system might opt for optimistic speculative execution and transaction commit/re-execution in case of conflicts to maximize performance

Thanks again for taking the time comment/answer
